src/Microsoft.OData.Client/ALinq/ResourceBinder.cs

```diff
@@ -797,6 +799,37 @@ internal static Expression ApplyOrdering(MethodCallExpression mce, bool descendi
             return input;
         }
 
+        private static Expression AnalyzeAggregation(MethodCallExpression methodCallExpr, AggregationMethod aggregationMethod)
+        {
+            Debug.Assert(methodCallExpr != null, "methodCallExpr != null");
+            if (methodCallExpr.Arguments.Count != 2)

```

Comment:
Why must the number of args be == 2?

Reply:
@odero Aggregation methods - `Average`, `Sum`, `Min`, `Max`) defined in `Queryable` and `Enumerable` classes have two sets of overloads for each aggregable type (`double`, `double?`, `float`, `float?`, `int`, `int?`, `long`, `long?`, `decimal`, `decimal?`)
- The first overload takes one parameter. Examples: `Average(IQueryable<Decimal>)` or `Average(IEnumerable<Decimal>)`.
To use this overload, you'd write the expression as follows:
    ```csharp
    context.Sales.Select(d1 => d1.Amount).Average(); // => Average(IEnumerable<Decimal>)
    ```
- The second overload takes two parameters. Example: `Average<TSource>(IQueryable<TSource>, Expression<Func<TSource,Decimal>>)`.
To use this overload, you'd write the expression as follows:
    ```csharp
    context.Sales.Average(d1 => d1.Amount);
    ```
**I implemented support for the second overload** in this pull request. That's the reason we're debug-asserting to verify that we're dealing with the second overload. It is succinct and requires the least amount of effort and almost no refactor to fit into the existing logic.

Implementing the first overload would be a lot of pain with little or no additional gain. Among other reasons, the `Select()` method is currently translated into `$select` expression. We'd need to do quite a bit of messy plumbing and refactor to handle `Select()` differently depending on whether or not its followed by an aggregation method, i.e.:
- Translate `Select(d1 => d1.Amount)` into `$select=Amount`, and
- Translate `Select(d1 => d1.Amount).Average()` into `$apply=aggregate(Amount with average as AverageAmount)`

The strategy would be to "look ahead" when doing the translation to determine which expression to translate into.
That support could be implemented at a later date. The effort would not be trivial so we'd need to have a strong justification.

We'd obviously need to document what is supported for the benefit of the users.